test/test_storage.py

```diff
@@ -0,0 +1,194 @@
+        content = {
+            'str': 'string',
+            'bytes': b'bytes',
+            'int': 1,
+            'float': 3.0,
+            'dict': {'a': 'b'},
+            'set': {'a', 'b'},
+            'list': [1, 2],
+        }
+        s = Sample(f, 'test', content)
+        handle = s.handle
+        f.save_snapshot(s)
+        del s
+        gc.collect()
+        res = f.load_snapshot(handle)
+        self.assertEqual(content, res.content)
+
+
+class TestSQLiteStorage(StoragePermutations, BaseTestCase):
+
+    def create_framework(self):
+        return framework.Framework(':memory:', None, None, None)
+
+
+class _MemoryStorageBackend:
+
+    def __init__(self):
+        self._values = {}
+        self._calls = []
+
+    def set(self, key, value):
+        self._calls.append(('set', key, value))
+        self._values[key] = value
+
+    def get(self, key):
+        self._calls.append(('get', key))
+        return self._values[key]
+
+    def delete(self, key):
+        self._calls.append(('delete', key))
+        self._calls.pop(key, None)
+
+
+class TestJujuStorage(StoragePermutations, BaseTestCase):
+
+    def create_framework(self):
+        mem_backend = _MemoryStorageBackend()
+        f = framework.Framework(':memory:', None, None, None)
+        f._storage = storage.JujuStorage(mem_backend)
+        return f
+
+
+class TestJujuStateBackend(BaseTestCase):
+
+    def test_is_not_available(self):
+        self.assertFalse(storage._JujuStorageBackend.is_available())
+
+    def test_is_available(self):
+        fake_script(self, 'state-get', 'echo ""')
+        self.assertTrue(storage._JujuStorageBackend.is_available())
+        self.assertEqual(fake_script_calls(self, clear=True), [])
+
+    def test_set_encodes_args(self):
+        t = tempfile.NamedTemporaryFile()
+        fake_script(self, 'state-set', dedent("""
+            cat >> {}
+            """).format(t.name))
+        backend = storage._JujuStorageBackend()
+        backend.set('key', {'foo': 2})
+        self.assertEqual(fake_script_calls(self, clear=True), [
+            ['state-set', '--file', '-'],
+        ])
+        t.seek(0)
+        content = t.read()
+        self.assertEqual(content.decode('utf-8'), dedent("""\
+            "key": |
+              {foo: 2}
+            """))
+
+    def test_get(self):
+        fake_script(self, 'state-get', dedent("""
+            echo 'foo: "bar"'
+            """))
+        backend = storage._JujuStorageBackend()
+        value = backend.get('key')
+        self.assertEqual(value, {'foo': 'bar'})
+        self.assertEqual(fake_script_calls(self, clear=True), [
+            ['state-get', 'key'],
+        ])
+
+    def test_set_and_get_complex_value(self):
+        t = tempfile.NamedTemporaryFile()
+        fake_script(self, 'state-set', dedent("""
+            cat >> {}
+            """).format(t.name))
+        backend = storage._JujuStorageBackend()
+        complex_val = {
```

Comment:
Are there any other types that we want to make sure to round-trip?
I'm not sure if I care about complex, but we could add it.
int
float
dict
set
str
bytes
list
tuple

are covered so far.

Reply:
`complex`, `bytearray`, `memoryview`, `frozenset`. I think I can live without all of these.

Reply:
We should restrict to only the types we care and test. If somebody needs `complex` (or whatever) in the future, we could add it. If we don't restrict it, in the future we will not be able to remove it (if we find a problem, for example) by fear of it being used in the wild.

Reply:
i think the current approach does just that :-)